File: backend/app/schemas/rfq.py
"""BatiConnect — Mise en concurrence encadree: RFQ schemas."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

# ---------------------------------------------------------------------------
# TenderRequest
# ---------------------------------------------------------------------------


class TenderRequestCreate(BaseModel):
    building_id: UUID
    title: str
    description: str | None = None
    work_type: str  # asbestos_removal | pcb_removal | lead_removal | hap_removal | radon_mitigation | pfas_remediation | multi_pollutant | other
    deadline_submission: datetime | None = None
    planned_start_date: date | None = None
    planned_end_date: date | None = None
    attachments_manual: list[str] | None = None


class TenderRequestUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    work_type: str | None = None
    deadline_submission: datetime | None = None
    planned_start_date: date | None = None
    planned_end_date: date | None = None
    attachments_manual: list[str] | None = None
    status: str | None = None


class TenderRequestRead(BaseModel):
    id: UUID
    building_id: UUID
    organization_id: UUID | None
    created_by_id: UUID
    title: str
    description: str | None
    scope_summary: str | None
    work_type: str
    deadline_submission: datetime | None
    planned_start_date: date | None
    planned_end_date: date | None
    status: str
    attachments_auto: list | None
    attachments_manual: list | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# TenderInvitation
# ---------------------------------------------------------------------------


class TenderInvitationCreate(BaseModel):
    contractor_org_ids: list[UUID]


class TenderInvitationRead(BaseModel):
    id: UUID
    tender_id: UUID
    contractor_org_id: UUID
    sent_at: datetime | None
    viewed_at: datetime | None
    responded_at: datetime | None
    status: str
    access_token: str | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# TenderQuote
# ---------------------------------------------------------------------------


class TenderQuoteCreate(BaseModel):
    invitation_id: UUID | None = None
    contractor_org_id: UUID
    total_amount_chf: Decimal | None = None
    currency: str = "CHF"
    scope_description: str | None = None
    exclusions: str | None = None
    inclusions: str | None = None
    estimated_duration_days: int | None = None
    validity_date: date | None = None
    document_id: UUID | None = None


class TenderQuoteRead(BaseModel):
    id: UUID
    tender_id: UUID
    invitation_id: UUID | None
    contractor_org_id: UUID
    total_amount_chf: Decimal | None
    currency: str
    scope_description: str | None
    exclusions: str | None
    inclusions: str | None
    estimated_duration_days: int | None
    validity_date: date | None
    document_id: UUID | None
    extracted_data: dict | None
    status: str
    submitted_at: datetime | None
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# TenderComparison
# ---------------------------------------------------------------------------


class TenderComparisonRead(BaseModel):
    id: UUID
    tender_id: UUID
    created_by_id: UUID
    comparison_data: dict | None
    selected_quote_id: UUID | None
    selection_reason: str | None
    attributed_at: datetime | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Attribution request body
# ---------------------------------------------------------------------------


class TenderAttributeRequest(BaseModel):
    quote_id: UUID
    reason: str | None = None

File: frontend/src/api/rfq.ts
import { apiClient } from '@/api/client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface TenderRequest {
  id: string;
  building_id: string;
  organization_id: string | null;
  created_by_id: string;
  title: string;
  description: string | null;
  scope_summary: string | null;
  work_type: string;
  deadline_submission: string | null;
  planned_start_date: string | null;
  planned_end_date: string | null;
  status: 'draft' | 'sent' | 'collecting' | 'closed' | 'attributed' | 'cancelled';
  attachments_auto: string[] | null;
  attachments_manual: string[] | null;
  created_at: string;
  updated_at: string;
}

export interface TenderInvitation {
  id: string;
  tender_id: string;
  contractor_org_id: string;
  sent_at: string | null;
  viewed_at: string | null;
  responded_at: string | null;
  status: string;
  access_token: string | null;
  created_at: string;
}

export interface TenderQuote {
  id: string;
  tender_id: string;
  invitation_id: string | null;
  contractor_org_id: string;
  total_amount_chf: number | null;
  currency: string;
  scope_description: string | null;
  exclusions: string | null;
  inclusions: string | null;
  estimated_duration_days: number | null;
  validity_date: string | null;
  document_id: string | null;
  extracted_data: Record<string, unknown> | null;
  status: string;
  submitted_at: string | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TenderComparisonEntry {
  quote_id: string;
  contractor_org_id: string;
  total_amount_chf: number | null;
  currency: string;
  estimated_duration_days: number | null;
  scope_description: string | null;
  inclusions: string | null;
  exclusions: string | null;
  validity_date: string | null;
  submitted_at: string | null;
  status: string;
}

export interface TenderComparisonStats {
  total_quotes: number;
  quotes_with_amount: number;
  amount_range_chf: { min: number | null; max: number | null };
  duration_range_days: { min: number | null; max: number | null };
}

export interface TenderComparisonData {
  entries: TenderComparisonEntry[];
  statistics: TenderComparisonStats;
  generated_at: string;
}

export interface TenderComparison {
  id: string;
  tender_id: string;
  created_by_id: string;
  comparison_data: TenderComparisonData | null;
  selected_quote_id: string | null;
  selection_reason: string | null;
  attributed_at: string | null;
  created_at: string;
}

export interface TenderCreatePayload {
  building_id: string;
  title: string;
  work_type: string;
  description?: string | null;
  deadline_submission?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  attachments_manual?: string[] | null;
}

export interface TenderUpdatePayload {
  title?: string;
  description?: string | null;
  work_type?: string;
  deadline_submission?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  attachments_manual?: string[] | null;
  status?: string;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export const rfqApi = {
  generateDraft: async (data: TenderCreatePayload): Promise<TenderRequest> => {
    const response = await apiClient.post<TenderRequest>('/rfq/generate', data);
    return response.data;
  },

  get: async (tenderId: string): Promise<TenderRequest> => {
    const response = await apiClient.get<TenderRequest>(`/rfq/${tenderId}`);
    return response.data;
  },

  update: async (tenderId: string, data: TenderUpdatePayload): Promise<TenderRequest> => {
    const response = await apiClient.put<TenderRequest>(`/rfq/${tenderId}`, data);
    return response.data;
  },

  send: async (tenderId: string, contractorOrgIds: string[]): Promise<TenderInvitation[]> => {
    const response = await apiClient.post<TenderInvitation[]>(`/rfq/${tenderId}/send`, {
      contractor_org_ids: contractorOrgIds,
    });
    return response.data;
  },

  submitQuote: async (
    tenderId: string,
    data: {
      contractor_org_id: string;
      invitation_id?: string;
      total_amount_chf?: number;
      scope_description?: string;
      exclusions?: string;
      inclusions?: string;
      estimated_duration_days?: number;
      validity_date?: string;
      document_id?: string;
    },
  ): Promise<TenderQuote> => {
    const response = await apiClient.post<TenderQuote>(`/rfq/${tenderId}/quotes`, data);
    return response.data;
  },

  extractQuoteData: async (tenderId: string, quoteId: string): Promise<TenderQuote> => {
    const response = await apiClient.post<TenderQuote>(`/rfq/${tenderId}/quotes/${quoteId}/extract`);
    return response.data;
  },

  generateComparison: async (tenderId: string): Promise<TenderComparison> => {
    const response = await apiClient.post<TenderComparison>(`/rfq/${tenderId}/compare`);
    return response.data;
  },

  attribute: async (tenderId: string, quoteId: string, reason?: string): Promise<TenderComparison> => {
    const response = await apiClient.post<TenderComparison>(`/rfq/${tenderId}/attribute`, {
      quote_id: quoteId,
      reason: reason ?? null,
    });
    return response.data;
  },

  listQuotes: async (tenderId: string): Promise<TenderQuote[]> => {
    const response = await apiClient.get<TenderQuote[]>(`/rfq/${tenderId}/quotes`);
    return response.data;
  },

  listByBuilding: async (buildingId: string): Promise<TenderRequest[]> => {
    const response = await apiClient.get<TenderRequest[]>(`/rfq/building/${buildingId}`);
    return response.data;
  },
};
